### PR TITLE
100% misc.autoupdate coverage

### PR DESCRIPTION
**Pull request by acogneau**

### CODE REVIEW

File: tests/unit/misc/test_autoupdate.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+from PyQt5.QtTest import QSignalSpy
+
+from qutebrowser.misc import autoupdate, httpclient
+
+
+class HTTPGetStub(httpclient.HTTPClient):
+    """A stub class for HTTPClient.
+
+        Attributes:
```
Comment (The-Compiler) [inline tests/unit/misc/test_autoupdate.py]: Minor thing, but please dedent this by one level:

```python
Attributes:
    foo: bar
```

File: tests/unit/misc/test_autoupdate.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    """
+
+    def __init__(self, success=True, json=None):
+        self._success = success
+        if json:
+            self._json = json
+        else:
+            self._json = '{"info": {"version": "test"}}'
+
+        super(HTTPGetStub, self).__init__()
```
Comment (The-Compiler) [inline tests/unit/misc/test_autoupdate.py]: You can/should use the Python 3 syntax here: `super().__init__()` - also I'd suggest moving the superclass init to the top of `__init__` so it gets executed first. It doesn't make a difference here, but it's the usual thing to do (tm) :wink:

File: tests/unit/misc/test_autoupdate.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    def get(self, url):
+        if self._success is True:
```
Comment (The-Compiler) [inline tests/unit/misc/test_autoupdate.py]: This is unidiomatic in Python - you'd just do `if self._success:`

File: tests/unit/misc/test_autoupdate.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+def test_get_version_success(qtbot):
+    """Test get_version() when success is emitted."""
+    http_stub = HTTPGetStub(success=True)
+    client = autoupdate.PyPIVersionClient(client=http_stub)
+
+    # Use a spy to inspect the signal
+    success_spy = QSignalSpy(client.success)
+    error_spy = QSignalSpy(client.error)
+
+    with qtbot.waitSignal(client.success, timeout=2000, raising=False):
```
Comment (The-Compiler) [inline tests/unit/misc/test_autoupdate.py]: You could drop the `success_spy` and use `raising=True` instead. I also think you can leave the timeout off, the default of 1s should be fine.

File: tests/unit/misc/test_autoupdate.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+def test_get_version_error(qtbot):
+    """Test get_version() when error is emitted."""
+    http_stub = HTTPGetStub(success=False)
+    client = autoupdate.PyPIVersionClient(client=http_stub)
+
+    # Use a spy to inspect the signal
+    success_spy = QSignalSpy(client.success)
+    error_spy = QSignalSpy(client.error)
+
+    with qtbot.waitSignal(client.error, timeout=2000, raising=False):
```
Comment (The-Compiler) [inline tests/unit/misc/test_autoupdate.py]: Same as above.

File: tests/unit/misc/test_autoupdate.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    """Test on_client_success() with invalid JSON."""
+    json = '{"invalid": { "json"}'
+    http_stub = HTTPGetStub(json=json)
+    client = autoupdate.PyPIVersionClient(client=http_stub)
+    error_spy = QSignalSpy(client.error)
+    client.get_version('test')
+    assert len(error_spy) == 1
+
+
+def test_invalid_keys():
```
Comment (The-Compiler) [inline tests/unit/misc/test_autoupdate.py]: This and `test_invalid_json` can probably be a parametrized test

**Comment by The-Compiler** [comment]
The errors you see are because I broke the tests yesterday and fixed it now - that was bad timing :laughing:

You can either ignore them, or rebase with `git rebase master` (after you updated master) and then `git push --force`

**Comment by acogneau** [comment]
@The-Compiler What errors? :^)

File: tests/unit/misc/test_autoupdate.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    def __init__(self, success=True, json=None):
+        super().__init__()
+        self._success = success
+        if json:
+            self._json = json
+        else:
+            self._json = '{"info": {"version": "test"}}'
+
+    def get(self, url):
```
Comment (The-Compiler) [inline tests/unit/misc/test_autoupdate.py]: Maybe also make sure the URL is correct here - something like:

```python
assert url == QUrl('https://pypi.python.org/pypi/test/json')
```

**Pull request merged**